Add steps and details for running 'unprivileged' Elastic Agent

kilfoyle opened the pull request:
These are ready for review now.

Main changes are in:  [Run Elastic Agent without administrative privileges](https://ingest-docs_bk_1087.docs-preview.app.elstc.co/guide/en/fleet/master/elastic-agent-unprivileged.html)

Smaller changes:
 - Added an `--unprivileged` option to the [elastic-agent install command](https://ingest-docs_bk_1087.docs-preview.app.elstc.co/guide/en/fleet/master/elastic-agent-cmd-options.html#_options_5). 
 - Added an [Error when running Elastic Agent commands with `sudo`](https://ingest-docs_bk_1087.docs-preview.app.elstc.co/guide/en/fleet/master/fleet-troubleshooting.html#agent-sudo-error) troubleshooting topic

Closes: #1077
Closes: https://github.com/elastic/elastic-agent/issues/4705
Closes: https://github.com/elastic/elastic-agent/issues/4653
Closes: https://github.com/elastic/elastic-agent/issues/4647
Closes: https://github.com/elastic/elastic-agent/issues/4675


Target: 8.15

ycombinator commented on docs/en/ingest-management/elastic-agent/elastic-agent-unprivileged-mode.asciidoc:
```diff
@@ -0,0 +1,125 @@
+[[unprivileged-running]]
+== Run {agent} in `unprivileged` mode
+
+TBD
```
@blakerouse Could you provide here the steps a user would need to take to install/run Agent in unprivileged mode please?  Thanks!

blakerouse replied:
All the commands remain the same except for `install` which has the new `--unprivileged` flag. Lets say in Fleet it says run `elastic-agent install --url <URL> --enrollment-token <TOKEN>` then you just need to add an additional `--unprivileged` flag to get it to install in unprivileged mode.

kilfoyle replied:
Here's what I added:

![Screenshot 2024-06-20 at 1 52 20 PM](https://github.com/elastic/ingest-docs/assets/41695641/e8197349-28a2-4630-916c-d294151ff4a9)

ycombinator commented on docs/en/ingest-management/elastic-agent/elastic-agent-unprivileged-mode.asciidoc:
```diff
@@ -0,0 +1,125 @@
+
+If an agent doesn't have the right level of privilege to read a data source, you can adjust the agent's privileges by adding the user running {agent} to the user group that has privileges to read the data source.
+
+TBD
```
@blakerouse Are there any specifics we could tell the user about here?

blakerouse replied:
The specifics are based on the operating system and integration. It's very hard to give specifics.

Just to provide context with `--unprivileged` mode 1 user and 1 group is created on the host. They names are the same across all operation systems.

`elastic-agent-user` is the user that is created and the service runs as.
`elastic-agent` group is created and any user in that group will have access to control and talk over the control protocol to the elastic agent daemon.

Let me give an example:

Say I install the Elastic Agent with `--unprivileged` then the installer will automatically create the `elastic-agent-user` and `elastic-agent` group. If I want say my user `blake` to be able to run `elastic-agent status` then I need to add the `blake` user to `elastic-agent` group. Once added to the group the `elastic-agent status` will work, until then a permission error with communicating with the control socket will occur.

kilfoyle replied:
Thanks! That's crystal clear. Here's what I've added:

![Screenshot 2024-06-20 at 2 34 37 PM](https://github.com/elastic/ingest-docs/assets/41695641/9af3a65f-1168-48df-8f69-3b53eea14035)

blakerouse commented on docs/en/ingest-management/elastic-agent/elastic-agent-unprivileged-mode.asciidoc:
```diff
@@ -0,0 +1,125 @@
+[[unprivileged-change-mode]]
+== Changing an {agent}'s privilege mode
+
+If an agent doesn't have the right level of privilege to read a data source, you can adjust the agent's privileges by adding the user running {agent} to the user group that has privileges to read the data source.
```
This is not correct. You need to add the `elastic-agent-user` to the group, not the running users name.

kilfoyle replied:
Changed to:

> If an agent doesn't have the right level of privilege to read a data source, you can adjust the agent's privileges by adding `elastic-agent-user` to the user group that has privileges to read the data source.

ycombinator commented:
Hi @kilfoyle, based on the comments in https://github.com/elastic/elastic-agent/issues/4889, https://github.com/elastic/elastic-agent/issues/4888, and https://github.com/elastic/elastic-agent/issues/4887, we should include a section somewhere in this documentation about running Elastic Agent sub-commands using the same user as the one Agent uses when installing itself in `--unprivileged` mode.  For specific suggestions on how to handle this scenario, see https://github.com/elastic/elastic-agent/issues/4889#issuecomment-2158635347.

kilfoyle commented:
> Hi @kilfoyle, based on the comments in https://github.com/elastic/elastic-agent/issues/4889, https://github.com/elastic/elastic-agent/issues/4888, and https://github.com/elastic/elastic-agent/issues/4887, we should include a section somewhere in this documentation about running Elastic Agent sub-commands using the same user as the one Agent uses when installing itself in --unprivileged mode. For specific suggestions on how to handle this scenario, see https://github.com/elastic/elastic-agent/issues/4889#issuecomment-2158635347.

Thanks @ycombinator! I've added a prominent warning about this in the "Run Elastic Agent in `unprivileged` mode` section:
![Screenshot 2024-06-21 at 12 51 57 PM](https://github.com/elastic/ingest-docs/assets/41695641/7e8b9f37-8353-4192-a294-889a05beb6b2)

---

And also a new section on the troubleshooting docs page:
![Screenshot 2024-06-21 at 12 51 33 PM](https://github.com/elastic/ingest-docs/assets/41695641/3da7dd6e-496d-4336-b28e-d60f60982eb6)

kilfoyle commented:
Just for my own reference, updated the Command reference install command syntax and added this description: 

![Screenshot 2024-06-21 at 1 27 47 PM](https://github.com/elastic/ingest-docs/assets/41695641/a7b08683-11c4-44b9-aa3d-ea4ef2721b3d)

blakerouse commented on docs/en/ingest-management/elastic-agent/elastic-agent-unprivileged-mode.asciidoc:
```diff
@@ -36,50 +36,89 @@ Doing so may result in <<agent-sudo-error,an error>> due to the agent not having
+
+|Run {agent} with the System integration
+|Log file error: `[elastic_agent.filebeat][error] Harvester could not be started on new file: /var/log/auth.log.1, Err: error setting up harvester: Harvester setup failed. Unexpected file opening error: Failed opening /var/log/auth.log.1: open /var/log/auth.log.1: permission denied`
+|Give read permission to the `elastic-agent` group for the `/var/log/syslog` and `/var/log/auth.log` files to fix this error.
```
I would actually recommend just adding the `elastic-agent-user` to the `adm` group on the host. This will allow the Elastic Agent to read the contents of that file without requiring changing the permissions of the file.

blakerouse commented on docs/en/ingest-management/elastic-agent/elastic-agent-unprivileged-mode.asciidoc:
```diff
@@ -36,50 +36,89 @@ Doing so may result in <<agent-sudo-error,an error>> due to the agent not having
+
+|Run {agent} with the System integration
+|On the `[Logs System] Syslog` dashboard, the `Syslog events by hostname`, `Syslog hostnames and processes` and `Syslog logs` visualizations are are missing data.
+|Give read permission to the `elastic-agent` group for the `/var/log/syslog` and `/var/log/auth.log` files to fix the missing visualizations.
```
I would actually recommend just adding the `elastic-agent-user` to the `adm` group on the host. This will allow the Elastic Agent to read the contents of that file without requiring changing the permissions of the file.

blakerouse commented on docs/en/ingest-management/elastic-agent/elastic-agent-unprivileged-mode.asciidoc:
```diff
@@ -36,50 +36,89 @@ Doing so may result in <<agent-sudo-error,an error>> due to the agent not having
+
+|Run {agent} with the System integration
+|Log file error: `failed to open Windows Event Log channel "Security": Access is denied`
+|Add the `elastic-agent` user to the `Event Log Users` group to fix this error.
```
```suggestion
|Add the `elastic-agent-user` user to the `Event Log Users` group to fix this error.
```

kilfoyle replied:
Fixed in [e4de01e](https://github.com/elastic/ingest-docs/pull/1087/commits/e4de01e1e369675b3901955d77607e6c0d0cdd47)

blakerouse commented on docs/en/ingest-management/elastic-agent/elastic-agent-unprivileged-mode.asciidoc:
```diff
@@ -36,50 +36,89 @@ Doing so may result in <<agent-sudo-error,an error>> due to the agent not having
-|`Exmaple error message...`
+|Run {agent} with the System integration
+|Most of the System and {agent} dashboard visualizations are missing all data.
+|Add the `elastic-agent` user to the `Event Log Users` group and update the permissions for the `HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\PartMgr` registry to fix the missing visualizations.
```
```suggestion
|Add the `elastic-agent-user` user to the `Event Log Users` group and update the permissions for the `HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\PartMgr` registry to fix the missing visualizations.
```

kilfoyle replied:
Fixed in [e4de01e](https://github.com/elastic/ingest-docs/pull/1087/commits/e4de01e1e369675b3901955d77607e6c0d0cdd47)

blakerouse commented on docs/en/ingest-management/elastic-agent/elastic-agent-unprivileged-mode.asciidoc:
```diff
@@ -36,50 +36,89 @@ Doing so may result in <<agent-sudo-error,an error>> due to the agent not having
+|Most of the System and {agent} dashboard visualizations are missing all data.
+|Add the `elastic-agent` user to the `Event Log Users` group and update the permissions for the `HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\PartMgr` registry to fix the missing visualizations.
+
+Note that the `elastic-user-agent` user may still not have access to all processes, so the lists in the `Top processes by CPU usage` and `Top processes by memory usage` visualizations may be incomplete. 
```
```suggestion
Note that the `elastic-agent-user` user may still not have access to all processes, so the lists in the `Top processes by CPU usage` and `Top processes by memory usage` visualizations may be incomplete. 
```

kilfoyle replied:
Fixed in [e4de01e](https://github.com/elastic/ingest-docs/pull/1087/commits/e4de01e1e369675b3901955d77607e6c0d0cdd47)

blakerouse commented on docs/en/ingest-management/elastic-agent/elastic-agent-unprivileged-mode.asciidoc:
```diff
@@ -36,50 +36,89 @@ Doing so may result in <<agent-sudo-error,an error>> due to the agent not having
+
+|Run {agent} with the System integration
+|On the `[Metrics System] Host overview` dashboard, the `Disk usage` visualizations are missing data. 
+|This occurs because direct access to the disk or a volume is restricted and not available to users without administrative privileges. Refer to link:https://learn.microsoft.com/en-us/windows/win32/secbp/running-with-special-privileges[Running with Special Privileges] in the Microsoft documentation for details.
```
I don't know exact which groups provide what but datadogs agent automatically adds the user to these groups allowing them access to a lot of information on Windows:

https://docs.datadoghq.com/agent/guide/windows-agent-ddagent-user/

Probably adding the `elastic-agent-user` to some of those groups might fix some issues here.

kilfoyle replied:
Here's what I've added:

![Screenshot 2024-06-24 at 10 16 04 AM](https://github.com/elastic/ingest-docs/assets/41695641/01eef7df-7a4b-44cc-a11a-9f73dfa53918)

@ycombinator, @VihasMakwana If possible it'd be really nice to test the newly added resolution.

ycombinator replied:
@VihasMakwana would you mind testing the resolution David added (see screenshot in previous comment) and posting your findings here?  Thanks.

VihasMakwana replied:
@blakerouse @ycombinator it turns out that datadog is using the Win32 PDH functions.
more about it here https://learn.microsoft.com/en-us/windows/win32/perfctrs/using-the-pdh-functions-to-consume-counter-data

----

@blakerouse even if i add the user to "Performance Monitor Group", it's not getting past [this section](https://github.com/elastic/elastic-agent-system-metrics/blob/main/metric/system/diskio/diskstat_windows_helper.go#L98-L104) (i.e. opening a volume) and it's throwing the error.

blakerouse replied:
@VihasMakwana dang it, was worth a shot.

kilfoyle replied:
Ah, too bad. Thanks a lot for testing it @VihasMakwana!

I've removed this:
>Adding the `elastic-agent-user` user to the `Performance Monitor Users` group, `Event Log Readers` group, and `Performance Log Users` group should resolve the missing data in the visualations.

ycombinator commented on docs/en/ingest-management/elastic-agent/elastic-agent-unprivileged-mode.asciidoc:
```diff
@@ -0,0 +1,185 @@
+
+|Run {agent} with the System integration
+|On the `[Metrics System] Host overview` dashboard, only the processes ran by the `elastic-agent-user` user are shown in the CPU and memory usage lists.
+|Give read permission to the `elastic-agent` group for the `/var/log/system.log` file to fix the missing visualizations.
```
I'm not sure this is the right resolution for this particular symptom. @blakerouse Please keep me honest but  I don't think there is any resolution here (short of running Agent as `root`, which we can't do in `unprivileged` mode by definition).

kilfoyle replied:
Ah, my mistake! Kaan's notes say only `This confirms that only the processes ran by the elastic-agent-user user are shown in the cpu and memory usage list.` And I see the resolution I added is for a different issue, specifically that on the `[Elastic Agent] Agent Info` dashboard,  the `[Elastic Agent] Most Active Agents` visualization isn't populated. I'll fix this.

ycombinator replied:
Not your mistake.  I went back and looked in https://github.com/elastic/elastic-agent/issues/4705 and it has the same incorrect resolutions too — I think it was just a copy/paste error. 🙂

blakerouse replied:
An option on macOS is to add the `elastic-agent-user` to the `admin` group. That is probably preferred from changing the ownership of that file, but possible that it gives the `elastic-agent-user` more permissions than the user wants.

kilfoyle replied:
I've changed this as shown:
![Screenshot 2024-06-25 at 11 05 42 AM](https://github.com/elastic/ingest-docs/assets/41695641/6edfbbd8-cc9b-4344-99a1-c6380542b37d)

ycombinator replied:
That looks really good to me, @kilfoyle!

ycombinator replied:
BTW, these same fixes need to be applied to the corresponding items for Linux.

kilfoyle replied:
I've updated three Linux `Run {agent} with the System integration` items as shown:

![Screenshot 2024-06-25 at 2 04 02 PM](https://github.com/elastic/ingest-docs/assets/41695641/614303b1-13b6-4bfd-a8f5-013fce921dc8)

---

The two Linux `Run {agent} and access the {agent} dashboards` items are unchanged:

![screen](https://github.com/elastic/ingest-docs/assets/41695641/cb46b077-aed2-4025-9d0b-4d58a11a2f67)

ycombinator commented on docs/en/ingest-management/elastic-agent/elastic-agent-unprivileged-mode.asciidoc:
```diff
@@ -0,0 +1,185 @@
+
+|Run {agent} and access the {agent} dashboards
+|On the `[Elastic Agent] Agents info` dashboard, visualizations including `Most Active Agents` and `Integrations per Agent` are missing data.
+|Give read permission to the `elastic-agent` group for the `/var/log/system.log` file to fix the missing visualizations.
```
I'm not sure this is the right resolution for this particular symptom.  @VihasMakwana would you mind confirming?

blakerouse replied:
Same `admin` group above applies.

kilfoyle replied:
Changed as shown:
![Screenshot 2024-06-25 at 11 05 42 AM](https://github.com/elastic/ingest-docs/assets/41695641/35673a0c-1993-4935-8249-61ae07698bc3)

ycombinator commented on docs/en/ingest-management/elastic-agent/elastic-agent-unprivileged-mode.asciidoc:
```diff
@@ -0,0 +1,185 @@
+
+|Run {agent} and access the {agent} dashboards
+|On the `[Elastic Agent] Integrations` dashboard, visualizations including `Integration Errors Table`, `Events per integration` and `Integration Errors` are missing data.
+|Give read permission to the `elastic-agent` group for the `/var/log/system.log` file to fix the missing visualizations.
```
I'm not sure this is the right resolution for this particular symptom.  @VihasMakwana would you mind confirming?

VihasMakwana replied:
@ycombinator I'll confirm and get back to you. I can't recall the resolution as @kaanyalti did the testing for this. I'll get back to you ASAP.

VihasMakwana replied:
@ycombinator as per @kaanyalti's note [here](https://github.com/elastic/elastic-agent/issues/4705#issuecomment-2155164637), this is the expected resolution.

I'll confirm it once and get back to you.

ycombinator replied:
Thanks @VihasMakwana.  I think what @kaanyalti posted there might've been a copy-paste typo.  It feels really odd to me that updating group ownership on the `/var/log/system.log` file would somehow affect the dashboards mentioned in this scenario.  I can't imagine that the data these dashboards are looking at is coming from that file.  So, yeah, appreciate you double checking this. Thanks again!

blakerouse replied:
Same as above, with `admin` group.

kilfoyle replied:
And this one similarly changed as shown:
![screen](https://github.com/elastic/ingest-docs/assets/41695641/0a5c54ba-c4ba-4e74-ade3-417cc5b204f2)

kilfoyle commented:
Thanks @ycombinator! It seems like we're near the finish line. Do you think there's anything else we should add or change?

ycombinator commented:
> Do you think there's anything else we should add or change?

Just https://github.com/elastic/ingest-docs/pull/1087#discussion_r1653265691
